### PR TITLE
Absolute Geolocation Error for CSLC-S1

### PR DESCRIPTION
**Pull request by vbrancat**
This PR aims at evaluating the Absolute Geolocation Error for OPERA CSLC-S1. 
The script `ale.py` requires as input:
1. A CSLC-S1 product
2. A CSV file containing the surveyed Corner Reflector locations (CR)

For each CR location contained in the CSLC-S1 product, the code:
1. Identifies the peak corresponding to the CR and computes the image coordinates
2. Computes ALE by differencing the computed CR location with the surveyed CR location

Results can be saved in a CSV file on disk or displayed on screen with a plot

### CODE REVIEW

File: src/compass/utils/ale.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        Array of Y-coordinates associated to CSLC-S1
+    x_spac: np.float
+        CSLC-S1 spacing along X-direction
+    y_spac: np.float
+        CSLC-S1 spacing along Y-direction
+    '''
+
+    with h5py.File(cslc_file, 'r') as h5:
+        x_vect = h5['science/SENTINEL1/CSLC/grids/x_coordinates'][()]
+        y_vect = h5['science/SENTINEL1/CSLC/grids/y_coordinates'][()]
```
Comment (seongsujeong) [inline src/compass/utils/ale.py:211]: Please note that there is half pixels of difference in the upper left corner's coordinates when the raster is loaded as `HDF5` and `NETCDF`. Based on what I've seen, the coordinates as NETCDF raster made more sense becuase the number was a multiple of 30.0

File: src/compass/utils/ale.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    cslc_poly: shapely.Polygon
+        Shapely polygon including CSLC-S1 valid values
+    '''
+    with h5py.File(cslc_file, 'r') as h5:
+        poly = h5['science/SENTINEL1/identification/bounding_polygon'][()]
+    cslc_poly = wkt.loads(poly)
+    return cslc_poly
+
+
+def get_clsc_epsg(cslc_file):
```
Comment (seongsujeong) [inline src/compass/utils/ale.py:267]:
```suggestion
def get_cslc_epsg(cslc_file):
```

File: src/compass/utils/ale.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    arr: np.ndarray
+        Array to use for identifying peak location
+    x_loc: np.int
+        First guess of peak location along X-coordinates
+    y_loc: np.int
+        First guess of peak location along Y-coordinates
+    ovs_factor: np.int
+        Oversampling factor
+    margin: int
+        Margin
```
Comment (seongsujeong) [inline src/compass/utils/ale.py:137]:
```suggestion
        half of the margin's size  in pixels
```

**Review by seongsujeong — COMMENTED** [review]
This is a good generalization of the code I've written [here](https://github.com/seongsujeong/compass_util/blob/main/compass_util.py). My initial comments are as below.

File: src/compass/utils/ale.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    Returns
+    -------
+    x_peak: np.float
+        Peak location along X-coordinates
+    y_peak: np.float
+        Peak location along Y-coordinate
+    '''
+
+    # Extract an area around x_loc, y_loc
+    img = arr[(y_loc - margin):(y_loc + margin),
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py:148]: It is an abbreviation for location. Do you want me to change the name of the variable if you think it is not sufficiently descriptive?

File: src/compass/utils/ale.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        err_str = f'{cslc_file} CSLC-S1 product does not exist'
+        raise FileNotFoundError(err_str)
+
+    # Check corner reflector file exists
+    if not os.path.exists(cr_file):
+        err_str = f'{cr_file} corner reflector file does not exist'
+        raise FileNotFoundError(err_str)
+
+    # Open corner reflector file
+    cr_df = pd.read_csv(cr_file)
```
[the remark this answers is not in the document]
Comment (scottstanie) [reply src/compass/utils/ale.py:46]: `df` is the typical variable name for a pandas `DataFrame`: https://pandas.pydata.org/docs/user_guide/10min.html

File: src/compass/utils/ale.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    cr_y = []
+    cr_x_cslc = []
+    cr_y_cslc = []
+
+    for idx, row in cr_df.iterrows():
+        cr_lat = row['Latitude (deg)']
+        cr_lon = row['Longitude (deg)']
+        cr_loc = geometry.Point(cr_lon, cr_lat)
+
+        if cslc_poly.contains(cr_loc):
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py:63]: Are you suggesting to apply a margin to the CR location?

File: src/compass/utils/ale.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py:46]: `df` stands for `DataFrame`. This is a typical variable abbreviation when using a `pandas` `DataFrame`

[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/utils/ale.py:46]: @scottstanie @vbrancat Good to know. Thanks for the information!

File: src/compass/utils/ale.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/utils/ale.py:63]: Yes. I think applying the margin in the image coordinate would be ideal. We can attempt to extract an image chip, and check if there are any "no value" in the chip. In that case we drop the image chip and the CR.

File: src/compass/utils/ale.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        File path to the CSLC-S1 product
+
+    Returns
+    -------
+    cslc_poly: shapely.Polygon
+        Shapely polygon including CSLC-S1 valid values
+    '''
+    with h5py.File(cslc_file, 'r') as h5:
+        poly = h5['science/SENTINEL1/identification/bounding_polygon'][()]
+    cslc_poly = wkt.loads(poly)
```
Comment (hfattahi) [inline src/compass/utils/ale.py:263]: you probably need to decode

```suggestion
    cslc_poly = wkt.loads(poly.decode('utf-8'))
```

File: src/compass/utils/ale.py [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+    # Compute absolute geolocation error along X and Y direction
+    cr_df['ALE_X'] = cr_df['CR_X_CSLC_PEAK'] - cr_df['CR_X']
+    cr_df['ALE_Y'] = cr_df['CR_Y_CSLC_PEAK'] - cr_df['CR_Y']
+
+    if csv_output_file is not None:
+        cr_df.to_csv(csv_output_file)
+
+    if plot_ale:
+        fig, ax = plt.subplots(figsize=(8, 6))
+        sc = ax.scatter(cr_df['ALE_Y'], cr_df['ALE_X'], s=200, alpha=0.8,
```
Comment (hfattahi) [inline src/compass/utils/ale.py:124]: X and Y are swapped?
```suggestion
        sc = ax.scatter(cr_df['ALE_X'], cr_df['ALE_Y'], s=200, alpha=0.8,
```

**Review by seongsujeong — COMMENTED** [review]
More comments are as below
 - Seongsu

**Comment by vbrancat** [comment]
@seongsujeong and @hfattahi the latest commits have:
1) Syncronized peak and geolocation error detection with Jeong's algorithm available here: https://github.com/seongsujeong/compass_util/blob/3aa086cdce903ce3853557592100563b69ceb05b/compass_util.py#L122. I have checked that the two codes produce the same geolocation error and they do. I have verified this on multiple bursts at multiple dates. Happy to point you to the data location
2) Expose more options in terms of polarization, margin, oversampling factor etc
3) Try to generalize the code to read NISAR GSLCs

Let me know your feedbacks

File: src/compass/utils/ale.py [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+        y_peak_vect.append(y_coord_vect[0] + y_peak * dy)
+
+    cr_df['CR_X_CSLC_PEAK'] = x_peak_vect
+    cr_df['CR_Y_CSLC_PEAK'] = y_peak_vect
+
+    # Compute absolute geolocation error along X and Y direction
+    cr_df['ALE_X'] = cr_df['CR_X_CSLC_PEAK'] - cr_df['CR_X']
+    cr_df['ALE_Y'] = cr_df['CR_Y_CSLC_PEAK'] - cr_df['CR_Y']
+
+    if csv_output_file is not None:
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py:143]: Done.

File: src/compass/utils/ale.py [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+                          help='OPERA L2 CSLC-S1 product')
+    required.add_argument('-c', '--cr-file', required=True, dest='cr_file',
+                          help='Corner reflector position CSV file '
+                               'same date as input CSLC-S1')
+    optional.add_argument('-s', '--save-csv', dest='save_csv', default=None,
+                          help='Save ALE in CSV file (default: None)')
+    optional.add_argument('-i', '--plot-ale', dest='plot_ale', default=False,
+                          help='Plot ALE')
+    optional.add_argument('-t', '--set', dest=set, default=False,
+                          help='If True, correct corner reflector positions for Solid Earth Tides')
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py]: Done.

File: src/compass/utils/ale.py [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+    required.add_argument('-p', '--cslc-s1', required=True, dest='cslc_file',
+                          help='OPERA L2 CSLC-S1 product')
+    required.add_argument('-c', '--cr-file', required=True, dest='cr_file',
+                          help='Corner reflector position CSV file '
+                               'same date as input CSLC-S1')
+    optional.add_argument('-s', '--save-csv', dest='save_csv', default=None,
+                          help='Save ALE in CSV file (default: None)')
+    optional.add_argument('-i', '--plot-ale', dest='plot_ale', default=False,
+                          help='Plot ALE')
+    optional.add_argument('-t', '--set', dest=set, default=False,
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py]: Resolved, thanks.

File: src/compass/utils/ale.py [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+    width = ds_in.RasterXSize
+    length = ds_in.RasterYSize
+
+    geo_trans = ds_in.GetGeoTransform()
+    x_spac = geo_trans[1]
+    y_spac = geo_trans[5]
+
+    # Generate x_vect and y_vect
+    x_vect = geo_trans[0] + np.arange(0, width)*x_spac
+    y_vect = geo_trans[1] + np.arange(0, length)*y_spac
```
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py]: Addressed, thanks.

File: src/compass/utils/ale.py (hunk 8 above)
Comment (vbrancat) [reply src/compass/utils/ale.py]: Resolved, thanks.

File: src/compass/utils/ale.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py:92]: This has been fixed in the latest commit.

[the remark this answers is not in the document]
Comment (vbrancat) [reply src/compass/utils/ale.py:92]: Not sure I totally agree with it. We are just checking that the CR position from the CSV is contained in the valid portion of the burst. In which case would the margin be useful? Like if the CR is at the boundary of a burst?

File: src/compass/utils/ale.py [hunk 13]
```diff
@@ -0,0 +1,536 @@
...
+
+import h5py
+import isce3
+import datetime as dt
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pysolid
+import shapely.wkt as wkt
+from osgeo import gdal
```
Comment (hfattahi) [inline src/compass/utils/ale.py:12]:
```suggestion
from osgeo import gdal, osr
```

File: src/compass/utils/ale.py [hunk 14]
```diff
@@ -0,0 +1,536 @@
+import argparse
+import os
+
+import h5py
+import isce3
+import datetime as dt
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pysolid
```
Comment (hfattahi) [inline src/compass/utils/ale.py:10]: I suggest to move this import to `correct_cr_tides` function:
```suggestion

```

File: src/compass/utils/ale.py [hunk 15]
```diff
@@ -0,0 +1,536 @@
...
+        cr_lat = row['Latitude (deg)']
+        cr_lon = row['Longitude (deg)']
+        cr_loc = geometry.Point(cr_lon, cr_lat)
+
+        # If the CR is contained in the geocoded SLC product
+        # get its position in the SLC image, otherwise, drop
+        # the CR from the pandas dataframe
+        if cslc_poly.contains(cr_loc):
+            # Convert corner lat/lon coordinates in UTM
+            cslc_epsg = get_cslc_epsg(cslc_file, mission_id=mission_id)
```
Comment (hfattahi) [inline src/compass/utils/ale.py:94]:
```suggestion
            cslc_epsg = get_cslc_epsg(cslc_file, mission_id=mission_id, pol=pol)
```

File: src/compass/utils/ale.py [hunk 16]
```diff
@@ -0,0 +1,536 @@
...
+                                                    sensing_stop,
+                                                    step_sec=5,
+                                                    display=False,
+                                                    verbose=False)
+    tide_e = np.mean(tide_e[0:2])
+    tide_n = np.mean(tide_n[0:2])
+
+    # Transform CR coordinates to UTM
+    cslc_epsg = get_cslc_epsg(cslc_file,
+                              mission_id=mission_id)
```
Comment (hfattahi) [inline src/compass/utils/ale.py:224]:
```suggestion
                              mission_id=mission_id, pol=pol)
```

File: src/compass/utils/ale.py [hunk 17]
```diff
@@ -0,0 +1,536 @@
...
+        Array of Y-coordinates associated to CSLC-S1
+    x_spac: np.float
+        CSLC-S1 spacing along X-direction
+    y_spac: np.float
+        CSLC-S1 spacing along Y-direction
+    '''
+    if mission_id == 'S1':
+        cslc_path = '/science/SENTINEL1/CSLC/grids/'
+    elif mission_id == 'NI':
+        cslc_path = '/science/LSAR/GSLC/grids/'
```
Comment (hfattahi) [inline src/compass/utils/ale.py:364]: Need to pass frequency for NISAR
```suggestion
        cslc_path = '/science/LSAR/GSLC/grids/frequency'
```

File: src/compass/utils/ale.py [hunk 18]
```diff
@@ -0,0 +1,536 @@
...
+
+    with h5py.File(cslc_file, 'r') as h5:
+        poly = h5[poly_path][()]
+
+    cslc_poly = wkt.loads(poly.decode('UTF-8'))
+
+    return cslc_poly
+
+
+def get_cslc_epsg(cslc_file, mission_id='S1'):
```
Comment (hfattahi) [inline src/compass/utils/ale.py:447]:
```suggestion
def get_cslc_epsg(cslc_file, mission_id='S1', pol='VV'):
```

File: src/compass/utils/ale.py [hunk 19]
```diff
@@ -0,0 +1,536 @@
...
+    Returns
+    -------
+    epsg: int
+        EPSG code identifying the projection of the
+        geocoded SLC product
+    '''
+    if mission_id == 'S1':
+        epsg_path = '/science/SENTINEL1/CSLC/grids/projection'
+    elif mission_id == 'NI':
+        epsg_path = '/science/LSAR/GSLC/frequencyA/projection'
```
Comment (hfattahi) [inline src/compass/utils/ale.py:469]:
```suggestion
        with h5py.File(cslc_file, 'r') as h5:
+            frequencies=h5["/science/LSAR/identification/listOfFrequencies"][()]
+            freq = frequencies[0].decode('utf-8')
+            frequency = f'frequency{freq}'
+
+        dataset_path = f'NETCDF:{cslc_file}://science/LSAR/GSLC/grids/{frequency}/{pol}'
+        ds = gdal.Open(dataset_path, gdal.GA_ReadOnly)
+        s = osr.SpatialReference(wkt=ds.GetProjection()).ExportToProj4()
+        crs = CRS.from_proj4(s)
+        epsg = crs.to_epsg()
```

File: src/compass/utils/ale.py [hunk 20]
```diff
@@ -0,0 +1,536 @@
...
+    else:
+        err_str = f'{mission_id} is not a valid mission identifier'
+        raise ValueError(err_str)
+
+    with h5py.File(cslc_file, 'r') as h5:
+        epsg = h5[epsg_path][()]
```
Comment (hfattahi) [inline src/compass/utils/ale.py:475]:
```suggestion
```

File: src/compass/utils/ale.py [hunk 21]
```diff
@@ -0,0 +1,536 @@
...
+    Returns
+    -------
+    x_tide_cr: float
+        Corner reflector position along X-direction corrected
+        for Solid Earth tides
+    y_tide_cr: float
+        Corner reflector position along Y-direction corrected
+        for Solid Earth tide
+    '''
+    # Get geocode SLC sensing start and stop
```
Comment (hfattahi) [inline src/compass/utils/ale.py:189]:
```suggestion
    import pysloid
    # Get geocode SLC sensing start and stop
```

**Review by hfattahi — COMMENTED** [review]
Few suggestions to make it work with NISAR GSLC

File: src/compass/utils/ale.py (hunk 15 above)
Comment (vbrancat) [reply src/compass/utils/ale.py:94]: Not sure why we need to specify the polarization here. We are reading the `epsg` from the `projection` dataset which should be common for every polarization

**Comment by vbrancat** [comment]
Thank you, @hfattahi! I addressed your comments. It would be great if I can have a toy GSLC so I can test my changes. the test on OPERA CSLC is pending and needs to be performed for the last set of changes

File: src/compass/utils/ale.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (seongsujeong) [reply src/compass/utils/ale.py:91]: >In which case would the margin be useful? Like if the CR is at the boundary of a burst?

If we don't have enough margin around the peak, then we can lose some signals around the peak in the SLC, which can affect the peak location in the oversampled SLC. In the case in the figure below, the first ovresampled SLC patch has peak at (2031, 2016), and the second one was at (2056, 2025)

output_20230410_1511

File: src/compass/utils/ale.py [hunk 22]
```diff
@@ -0,0 +1,551 @@
...
+        ax.axhline(0, color='black')
+        ax.axvline(0, color='black')
+        ax.set_xlabel('Easting error (m)')
+        ax.set_ylabel('Northing error (m)')
+        fig.suptitle('Absolute geolocation error (AGE)')
+        plt.show()
+
+
+def correct_cr_tides(cslc_file, cr_lat, cr_lon,
+                     mission_id='S1', pol=pol):
```
Comment (hfattahi) [inline src/compass/utils/ale.py:166]:
```suggestion
                     mission_id='S1', pol='VV'):
```

**Review by seongsujeong — COMMENTED** [review]
@vbrancat Thanks for your work at this PR. As I mentioned in the comment above, the result from this PR looks consistent to my code. I've left couple of suggestions in this iteration.

One more suggestion is to add this `ale.py` to `setup.cfg` in case you want this code be installed as a script when installing COMPASS.

File: src/compass/utils/ale.py [hunk 23]
```diff
@@ -0,0 +1,551 @@
...
+        Array of Y-coordinates associated to CSLC-S1
+    x_spac: np.float
+        CSLC-S1 spacing along X-direction
+    y_spac: np.float
+        CSLC-S1 spacing along Y-direction
+    '''
+    if mission_id == 'S1':
+        cslc_path = '/science/SENTINEL1/CSLC/grids/'
+    elif mission_id == 'NI':
+        cslc_path = '/science/LSAR/GSLC/grids/frequencyA'
```
Comment (hfattahi) [inline src/compass/utils/ale.py:370]:
```suggestion
        cslc_path = '/science/LSAR/GSLC/grids/frequencyA/'
```

**Review by hfattahi — APPROVED** [review]
Thanks @vbrancat for making the changes to make the script work with NISAR GSLC. I have to fix couple of things on NISAR RSLC and GSLC itself before further testing. But that is beyond this PR and should not hold this PR. The current PR looks good to me. Thank you for your nice work.

**Comment by seongsujeong** [comment]
> @seongsujeong not sure why I am not able to comment on your posted comment, do you think that a fixed margin would do the trick? If that is the case, you much would you recommend?

I think fixed margin would be okay as long as we set up enough margins around the peak. In my pessimistic estimate I would say about 6m in geolocation, about 3 pixels in peak radius, (so, 3*sqrt(5*10)=21.2), and 1.5 of safe factor. That would be about 33m, so I would suggest `35m` of margins.

**Comment by vbrancat** [comment]
@seongsujeong all your comments have been addressed :D

**Review by seongsujeong — APPROVED** [review]
Thanks for addressing my comments and suggestions. The PR looks good to me. Nice work!

**Pull request merged**